### PR TITLE
CStore switch FxHashMap to IndexVec

### LINKED ISSUE #46876: compiler code cleanup: replace metas: HashMap with IndexVec
While investigating the root cause(s) of #46112 and discussing (in the `#rustc` IRC chat room) the use of a hashmap mapping crate-numbers to crate-metadata, @eddyb [at some point exclaimed](https://botbot.me/mozilla/rustc/2017-12-19/?msg=94834353&page=1) "why is that not an `IndexVec`?"

One word: Legacy. Big time: the `metas` field has been a hashmap ever since `metadata::cstore` was introduced, 6.5 years ago in b23ecd47ceb7468d5132b9bf81d1d9c4e3fd1084 ([relevant line in diff](https://github.com/rust-lang/rust/commit/b23ecd47ceb7468d5132b9bf81d1d9c4e3fd1084#diff-b3412c376b0c44ca73fd5f1aa7db9193R5))

It should be a useful (and hopefully simple?) exercise to replace the hashmap that is [now there](https://github.com/rust-lang/rust/blob/970c613e4a785325469d4f694a16505d5deac17a/src/librustc_metadata/cstore.rs#L94) with an `IndexVec`.

### PR DESCRIPTION
**Pull request by Eh2406**
This is a first attempt to fix #46876.

### CODE REVIEW

**Comment by arielb1** [comment]
Your code does not compile:
```
[00:15:32] error[E0507]: cannot move out of borrowed content
[00:15:32]    --> /checkout/src/librustc_metadata/cstore.rs:114:9
[00:15:32]     |
[00:15:32] 114 |         self.metas.borrow().get(cnum).unwrap().unwrap().clone()
[00:15:32]     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot move out of borrowed content
[00:15:32] 
[00:15:32] error[E0596]: cannot borrow immutable local variable `met` as mutable
[00:15:32]    --> /checkout/src/librustc_metadata/cstore.rs:121:13
[00:15:32]     |
[00:15:32] 119 |         let met = self.metas.borrow_mut();
[00:15:32]     |             --- consider changing this to `mut met`
[00:15:32] 120 |         while met.len() <= cnum.index() {
[00:15:32] 121 |             met.push(None);
[00:15:32]     |             ^^^ cannot borrow mutably
[00:15:32] 
[00:15:32] error[E0596]: cannot borrow immutable local variable `met` as mutable
[00:15:32]    --> /checkout/src/librustc_metadata/cstore.rs:123:9
[00:15:32]     |
[00:15:32] 119 |         let met = self.metas.borrow_mut();
[00:15:32]     |             --- consider changing this to `mut met`
[00:15:32] ...
[00:15:32] 123 |         met[cnum] = Some(data);
[00:15:32]     |         ^^^ cannot borrow mutably
[00:15:32] 
```

File: src/librustc_metadata/cstore.rs [hunk 1]
```diff
@@ -111,18 +111,25 @@ impl CStore {
     }
 
     pub fn get_crate_data(&self, cnum: CrateNum) -> Rc<CrateMetadata> {
-        self.metas.borrow().get(&cnum).unwrap().clone()
+        self.metas.borrow().get(cnum).unwrap().unwrap().clone()
```
Comment (arielb1) [inline src/librustc_metadata/cstore.rs]: Use `self.metas.borrow()[cnum].clone().unwrap()`

File: src/librustc_metadata/cstore.rs [hunk 2]
```diff
@@ -111,18 +111,25 @@ impl CStore {
...
     }
 
     pub fn set_crate_data(&self, cnum: CrateNum, data: Rc<CrateMetadata>) {
-        self.metas.borrow_mut().insert(cnum, data);
+        use rustc_data_structures::indexed_vec::Idx;
+        let met = self.metas.borrow_mut();
```
Comment (arielb1) [inline src/librustc_metadata/cstore.rs]: This needs to be `let mut met`

**Comment by Eh2406** [comment]
Thanks for the heads up.  I got distracted trying to test locally.

**Comment by Eh2406** [comment]
Ok, I am confused, what does that test fail have to do with my changes?

**Comment by pnkfelix** [comment]
@Eh2406 Assuming that you are referring to the change to the output for ui/print_type_sizes/niche-filling.rs ... here is my current understanding...

I think we are seeing some (currently undiagnosed) non-deterministic treatment of `std::foo::bar` vs `core::foo::bar` , and your change is causing a ripple effect there.

You can see some recent notes from me on this at #46905, in particular the commit message for 5f85764788dba12d57a8955619be53634386228a

Note in particular that that tests *originally* said `core::cmp::Ordering`; it was only changed to `std::cmp::Ordering` very recently, in #46708 (which I am becoming increasingly concerned about and am considering reverting in favor of  #46838  ...).
 * Of course the *ideal* thing would be for someone to explain why it switched from `core` to `std` in the first place as part of #46708 ; I had thought it was obviously better at the time, but I have since come to understand the point of the comment noted in the log message for 5f85764788dba12d57a8955619be53634386228a : 
```rust
    Entry::Occupied(mut entry) => {
        // If `child` is defined in crate `cnum`, ensure
        // that it is mapped to a parent in `cnum`.
        if child.krate == cnum && entry.get().krate != cnum {
            entry.insert(parent);
        }
    }
```
 * namely, I think the idea of that code is that its better to prefer the path that corresponds to an items *definition*, if that path is public, rather than a path of a mere re-export of the item.

----

In the meantime I personally would be fine with you just updating the print-type-sizes.stdout to match the new output. (Which after all is the same as the old output...)

**Comment by Eh2406** [comment]
Ok so I switched from `core::cmp::Ordering` to `std::cmp::Ordering` in the test, we will see if that fixes things.

**Comment by Eh2406** [comment]
Any ideas what is up with these errors? Why would this change lead to differences in the error messages?

**Comment by Eh2406** [comment]
Is there any Ideas on how to proceed? Should I try a rebase?

**Comment by Eh2406** [comment]
All the PRs referenced by @pnkfelix have been merged, so I tried a rebase (and I squashed while I was doing). We will see if it works.

**Comment by Eh2406** [comment]
Ping. The test got fixed by other prs. So this is now back to being a small simple change, with green CI.

**Comment by Eh2406** [comment]
Time out on appveyor. So did the last few commits. So we are on hold for that to get straightened out, yes?

**Pull request merged**